syntax = "proto3";

/**
Data structures
 */
message Ballot {
    int32 id = 1;
    int32 ballotNum = 2;
}
message PValue {
    Ballot ballot = 1;
    string client = 2;
    string payload = 3;
}
message Batch {
    string client = 1;
    string request = 2;
}

/**
Official Paxos protocol messages: p1a, p1b, p2a, p2b
 */
message ProposerToAcceptor {
    enum Type {
        p1a = 0;
        p2a = 1;
    }

    int32 messageId = 1;
    Type type = 2;
    string acceptorGroupId = 3; //Acceptor group ID of recipient
    Ballot ballot = 4;
    int32 slot = 5;
    string client = 6;
    string payload = 7;
}

message AcceptorToProxyLeader {
    enum Type {
        p1b = 0;
        p2b = 1;
    }

    int32 messageId = 1;
    Type type = 2;
    string acceptorGroupId = 3;
    Ballot ballot = 4;
    int32 slot = 5;
    map<int32, PValue> log = 6;
}

message ProxyLeaderToProposer {
    enum Type {
        p1b = 0;
        p2b = 1;
        heartbeat = 2;
    }

    int32 messageId = 1;
    Type type = 2;
    string acceptorGroupId = 3;
    Ballot ballot = 4;
    int32 slot = 5;
    map<int32, string> committedLog = 6;
    map<int32, PValue> uncommittedLog = 7;
    repeated string connectedAcceptorGroupIds = 8;
}